### PR TITLE
Cinematic Foundational Systems Improved

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
This PR breaks the functionality of Cinematics, as this PR is an improved first half of this other draft PR: https://github.com/codecombat/codecombat/pull/5273 (which you shouldn't look at).

Don't worry too much that cinematics are broken. It'll be fixed in the next PR and it's feature gated.
Instead this PR locks down some foundational systems with tests.

Another reason I split out this piece is because adding tests really made the code review huge!

## Features

- Add selectors and their tests for all the features coming next. For fearless selecting into the Cinematic data!
- Add two higher order commands and tests.
- Test `parseShot` to ensure its functionality is valid.

- Refactor commands into a single utility file `commands.js`.
- Refactor cinematic schema to only include commands that have implementations.

### CODE REVIEW

File: app/schemas/selectors/cinematic.js [hunk 1]
```diff
@@ -30,10 +30,14 @@
  */
 const compose = (...fns) => initial => fns.reduce((v, fn) => fn(v), initial)
 
+/**
+ * @typedef TypeThangTypeSlug
```
Comment (shubhi1092) [inline app/schemas/selectors/cinematic.js]: Should it be `{Object}` ?

File: app/schemas/models/cinematic.schema.js [hunk 2]
```diff
@@ -1,125 +1,99 @@
...
     }),
-    clearObjects: c.object({
-      title: 'Clear Object',
+    clearBackgroundObject: c.object({
+      title: 'Clear Background Object',
       description: 'Clears the background objects from the screen after a given duration'
     }, {
       triggerStart: c.int({ title: 'Trigger Start(ms)', description: 'The number of milliseconds until background object is cleared' })
-    }),
-    animationTrigger: c.object({
```
Comment (shubhi1092) [inline app/schemas/models/cinematic.schema.js]: Why are we removing a bunch of properties from the schema?

File: app/schemas/selectors/cinematic.js [hunk 3]
```diff
@@ -93,6 +106,79 @@ const leftCharacter = shotSetup => (shotSetup || {}).leftThangType
...
+ * @returns {string|undefined}
+ */
+const slug = o => (o || {}).slug
+
+/**
+ * Returns properties required to place a background Lank.
+ * @param {Object} backgroundArt
+ * @returns {Object|undefined} a background object
+ */
+const background = backgroundArt => {
```
Comment (shubhi1092) [inline app/schemas/selectors/cinematic.js]: Naming suggestion: `getBackgroundFromBackgroundArt`

File: app/schemas/selectors/cinematic.js [hunk 4]
```diff
@@ -93,6 +106,79 @@ const leftCharacter = shotSetup => (shotSetup || {}).leftThangType
...
+  }
+}
+
+/**
+ * Check if this is the hero type.
+ * If so returns the properties associated.
+ * The caller must get the thangType original using:
+ * `me.get('heroConfig').thangType`
+ * @param {Object|undefined} thangType
+ * @returns {Object|undefined} the thang and hero original.
```
Comment (shubhi1092) [inline app/schemas/selectors/cinematic.js]: It looks like we are only returning thang properties like pos, scaleX, scaleY, and not original

File: app/schemas/selectors/cinematic.js [hunk 5]
```diff
@@ -102,49 +188,136 @@ const characterThangTypeSlug = character => {
...
+  const thangType = getCharacterFromThangTypeSchema((character || {}).thangType)
+  if (!thangType) {
     return
   }
 
-  if (!character.slug) {
-    throw new Error(`no slug on char`)
+  const { slug, thang } = thangType
+
+  if (typeof character.enterOnStart !== 'boolean') {
```
Comment (shubhi1092) [inline app/schemas/selectors/cinematic.js]: Would it not always be boolean since we have it enforced with the schema?

File: app/schemas/selectors/cinematic.js (hunk 3 above)
Comment (shubhi1092) [reply app/schemas/selectors/cinematic.js]: I see a naming pattern in this file, sometimes we are using `get` with the name of the function and sometimes not. Feel free to ignore my naming suggestion if it falls under the pattern, but I would like to understand when are we using `get` vs not.

File: app/views/play/cinematic/Command/AbstractCommand.js [hunk 6]
```diff
@@ -98,60 +98,3 @@ export default class AbstractCommand {
     return Promise.resolve()
   }
 }
-
```
Comment (shubhi1092) [inline app/views/play/cinematic/Command/AbstractCommand.js]: Why are we removing all this stuff from here?

Comment (shubhi1092) [reply app/views/play/cinematic/Command/AbstractCommand.js]: Oh I see now that it's in another file below (ignore the comment here)

File: app/views/play/cinematic/Command/commands.js [hunk 7]
```diff
@@ -40,3 +41,121 @@ export class Sleep extends AbstractCommand {
...
+    }
+    this.cancel = () => {}
+  }
+}
+
+export class Noop extends AbstractCommand {
+  run () {
+    return Promise.resolve()
+  }
+}
```
Comment (shubhi1092) [inline app/views/play/cinematic/Command/commands.js]: Are there any changes in the code upto this or is it moved from AbstractCommand to here as it is?

File: test/app/lib/cinematic/cinematicParser.spec.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    const shot = {
+      setupShot: 'Example setup shot'
+    }
+    const results = parseShot(shot, systems)
+    expect(systems[0].parseSetupShot).toHaveBeenCalledWith({ setupShot: 'Example setup shot' })
+    expect(systems[1].parseSetupShot).toHaveBeenCalledWith({ setupShot: 'Example setup shot' })
+    expect(results).toEqual([['setup command1', 'setup command2']])
+  })
+
+  it('parses nothing correctly', () => {
```
Comment (shubhi1092) [inline test/app/lib/cinematic/cinematicParser.spec.js]: The description of the test looks different than the test.

File: app/views/play/cinematic/Command/commands.js [hunk 9]
```diff
@@ -40,3 +41,121 @@ export class Sleep extends AbstractCommand {
...
+export class SequentialCommands extends AbstractCommand {
+  /**
+   * @param {AbstractCommand[]} commands - List of commands to run sequentially.
+   */
+  constructor (commands) {
+    super()
+    this.commandRunner = new CommandRunner(commands)
+  }
+
+  [run] () {
```
Comment (shubhi1092) [inline app/views/play/cinematic/Command/commands.js]: I am curious what does it mean to write function name in square brackets `[]`

File: test/app/lib/cinematic/AbstractCommand.spec.js [hunk 10]
```diff
@@ -29,23 +41,48 @@ describe('AbstractCommand', () => {
...
+    for (const fn of simpleRunFunctions) {
       expect(() => (new PassInRunCommand(fn))[run]()).toThrow()
     }
   })
 })
 
+describe('SyncFunction', () => {
+  it('runs any function passed in', () => {
+    for (const fn of simpleRunFunctions) {
+      expect(() => (new SyncFunction(fn))[run]()).not.toThrow()
```
Comment (shubhi1092) [inline test/app/lib/cinematic/AbstractCommand.spec.js]: Should we also check if `run` is called here?

File: test/app/lib/cinematic/AbstractCommand.spec.js [hunk 11]
```diff
@@ -29,23 +41,48 @@ describe('AbstractCommand', () => {
...
+      },
+      done
+    ]
+
+    const commands2 = commands.map(f => new SyncFunction(f));
+
+    (new SequentialCommands(commands2))[run]()
+  })
+})
+
```
Comment (shubhi1092) [inline test/app/lib/cinematic/AbstractCommand.spec.js]: Do we also need tests for ConcurrentCommands?

File: test/app/lib/cinematic/cinematicParser.spec.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    const systems = [{
+      parseDialogNode: jasmine.createSpy().and.returnValue(['dialog commands'])
+    }]
+    const shot = {
+      dialogNodes: ['DialogNode1', 'DialogNode2']
+    }
+    const results = parseShot(shot, systems)
+    expect(systems[0].parseDialogNode.calls.all().map(o => o.args)).toEqual([['DialogNode1'], ['DialogNode2']])
+    // I don't like this test but I couldn't figure out another way.
+    expect(JSON.stringify(results)).toEqual('[[{"commands":["dialog commands"]}],[{"commands":["dialog commands"]}]]')
```
Comment (shubhi1092) [inline test/app/lib/cinematic/cinematicParser.spec.js]: I think this is fine, otherwise we could alo match individual entries in the array returned. But this also looks fine.

File: app/views/play/cinematic/Command/commands.js (hunk 9 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/commands.js]: Not sure of the exact name. But it works out the expression for the method.
[Small demo](https://repl.it/@AndrewJakubowic/Example-of-class-methods)

Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/commands.js]: You can do a similar thing with object literals.

```
const name = 'Andrew'

const person = {
 [name]: 'blah'
}

// => { Andrew: 'blah }
```

File: test/app/lib/cinematic/AbstractCommand.spec.js (hunk 11 above)
Comment (AndrewJakubowicz) [reply test/app/lib/cinematic/AbstractCommand.spec.js]: I know you reviewed well because you saw this omission.
I have no idea how to test ConcurrentCommands. It's like a concurrent nightmare I have sometimes.

I would love to discuss what you think are good features to test. A simpler example of the same concept is how would you test `Promise.all`.
I will think about this because I think this is a necessary test.

File: app/views/play/cinematic/Command/commands.js (hunk 9 above)
Comment (shubhi1092) [reply app/views/play/cinematic/Command/commands.js]: Nice, thanks for the examples.

File: test/app/lib/cinematic/AbstractCommand.spec.js (hunk 11 above)
Comment (shubhi1092) [reply test/app/lib/cinematic/AbstractCommand.spec.js]: Yeah it seems difficult to test. I think the basic thing that we can test for now is to check that it runs the commands and resolves if every command is executed successfully, or gives an error if anyone fails. Does that sound right?

File: app/views/play/cinematic/Command/commands.js (hunk 9 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/commands.js]: Added this to the design slab: https://codecombat.slab.com/posts/cinematic-system-design-2cg8x7si#internal-details-abstract-command

File: app/schemas/models/cinematic.schema.js (hunk 2 above)
Comment (AndrewJakubowicz) [reply app/schemas/models/cinematic.schema.js]: We add them back later. They were just adding bloat for this PR.

File: app/schemas/selectors/cinematic.js (hunk 3 above)
Comment (AndrewJakubowicz) [reply app/schemas/selectors/cinematic.js]: tl;dr intention  was  `get` prefix means public. Otherwise use property name.

I think the file could  be organized to make the naming more clear. Or maybe alternative names should be used.
Basically private getters are named by their property names. Whilst the public getter from the `Shot` or  `DialogNode` has the `get` prefix.

With that intention in mind is there a better way to name the private methods? I copied a functional pattern where the private functions can be composed as if using the `.` operator.

E.g.
`shot.shotSetup.leftCharacter` is the same as `compose(shotSetup, leftCharacter)(shot)`.
Thus the functions reflect the attributes (with the additional verification and chaining `undefined`).

File: app/schemas/selectors/cinematic.js (hunk 4 above)
Comment (AndrewJakubowicz) [reply app/schemas/selectors/cinematic.js]: Good call out. I wonder if JsDoc lets me document the result in the way I saw you documenting the anonymous objects in your excellent PR?

File: app/schemas/selectors/cinematic.js (hunk 5 above)
Comment (AndrewJakubowicz) [reply app/schemas/selectors/cinematic.js]: That's true.  I'll remove this check.

File: app/views/play/cinematic/Command/commands.js (hunk 7 above)
Comment (AndrewJakubowicz) [reply app/views/play/cinematic/Command/commands.js]: The `SequentialCommands` and `ConcurrentCommands` below are new. The `AnimeCommand`  is just moved. Sorry for not making that clearer.

File: test/app/lib/cinematic/AbstractCommand.spec.js (hunk 10 above)
Comment (AndrewJakubowicz) [reply test/app/lib/cinematic/AbstractCommand.spec.js]: That's a great additional test. I'll add that! Thanks!

File: test/app/lib/cinematic/AbstractCommand.spec.js (hunk 11 above)
Comment (AndrewJakubowicz) [reply test/app/lib/cinematic/AbstractCommand.spec.js]: That does sound right. I'll add the case that they all run.
I'll also add the  case that they all cancel.
I can write an error test. However the error should just bubble up and throw as we don't catch them.

Thanks for the excellent suggestions.

File: test/app/lib/cinematic/cinematicParser.spec.js (hunk 8 above)
Comment (AndrewJakubowicz) [reply test/app/lib/cinematic/cinematicParser.spec.js]: I have no idea where that description came from.
It should definitely be something like `dialogNode array data passed into parseDialogNode system`.

Great pickup!

File: test/app/lib/cinematic/cinematicParser.spec.js (hunk 12 above)
Comment (AndrewJakubowicz) [reply test/app/lib/cinematic/cinematicParser.spec.js]: The stringified version works because the `new Command` that is wrapped around the  dialog node gets turned into the `"commands"` property in that object literal. I wonder if all JavaScript classes turn into an object literal when stringified?
Do you know how we could check equality between two  command objects or: Command == Command in Js?

File: test/app/models/Cinematic.spec.js [hunk 13]
```diff
@@ -1,5 +1,19 @@
...
+  getRightCharacterThangTypeSlug,
+  getLeftHero,
+  getRightHero,
+  getClearBackgroundObject,
+  getBackgroundObject,
+  getBackground,
+  getClearText,
+  getSpeaker,
+  getBackgroundSlug,
+  exitCharacter,
```
Comment (AndrewJakubowicz) [inline test/app/models/Cinematic.spec.js]: @shubhi1092 I responded above, saying that I was only exporting the `get<selector>` but then I broke my own naming rules here!!! :<
Will fix

File: app/schemas/selectors/cinematic.js (hunk 3 above)
Comment (shubhi1092) [reply app/schemas/selectors/cinematic.js]: Thanks for clarifying the naming pattern used in this file. I dont think we need to change that pattern, but reorganizing/adding a comment about it could help make it clearer.

File: app/schemas/selectors/cinematic.js (hunk 4 above)
Comment (shubhi1092) [reply app/schemas/selectors/cinematic.js]: Can you tell me more about which code you are referring to from my PR? (I am not able to recall)

File: test/app/lib/cinematic/cinematicParser.spec.js (hunk 12 above)
Comment (shubhi1092) [reply test/app/lib/cinematic/cinematicParser.spec.js]: Not sure, but I wonder if we can check property wise, like `results.commands` to be equal to `expectedResult.commands`?

File: app/schemas/selectors/cinematic.js (hunk 3 above)
Comment (AndrewJakubowicz) [reply app/schemas/selectors/cinematic.js]: I'll document this somewhere and re-organizing selectors as a post merge task.

File: app/schemas/selectors/cinematic.js (hunk 4 above)
Comment (AndrewJakubowicz) [reply app/schemas/selectors/cinematic.js]: [This pattern](https://stackoverflow.com/a/6460748) that unfortunately doesn't have a return version!

Comment (AndrewJakubowicz) [reply app/schemas/selectors/cinematic.js]: Also fixing the documentation as you suggested :)

Comment (shubhi1092) [reply app/schemas/selectors/cinematic.js]: Oh yeah, it doesnt have a return version. Maybe we can use `typedef` to define properties for return objects if required.
